src/common/RedBlackTree.mo

```diff
@@ -0,0 +1,366 @@
+
+  public func rev() : I.Iter<(X, Y)> = toIter(tree, #r2l);
+
+};
+
+
+type IterRep<X, Y> = List.List<{#tr:Tree<X, Y>; #xy:(X, ?Y)}>;
+
+public func toIter<X, Y>(t:Tree<X, Y>, dir:{#l2r; #r2l}) : I.Iter<(X, Y)> {
+  object {
+    var trees : IterRep<X, Y> = ?(#tr t, null);
+    public func next() : ?(X, Y) {
+      switch (dir, trees) {
+      case (_, null) { null };
+      case (_, ?(#tr(#leaf), ts))                     { trees := ts; next() };
+      case (_, ?(#xy xy, ts))                         { trees := ts; switch (xy.1) { case null next(); case (?y) ?(xy.0, y) } };
+      case (#l2r, ?(#tr(#node(_, l, xy, r)), ts))     { trees := ?(#tr l, ?(#xy xy, ?(#tr r, ts))); next() };
+      case (#r2l, ?(#tr(#node(_, l, xy, r)), ts))     { trees := ?(#tr r, ?(#xy xy, ?(#tr l, ts))); next() };
+      }
+    };
+  }
+};
+
+
+/*
+Triggers compiler Error:
+  Fatal error: exception "Assert_failure mo_frontend/coverage.ml:300:4"
+
+public func height<X, Y>(t:Tree<X, Y>) : Nat {
+  switch t {
+    case (#leaf) 0;
+    case (#node(_, _, l, r)) {
+           ExtNat.max(height(l), height(r))
+         }
+  }
+};
+*/
+
+
+/*
+Triggers compiler error:
+  Fatal error: exception "Assert_failure mo_frontend/coverage.ml:300:4"
+
+public func size<X, Y>(t:Tree<X, Y>) : Nat {
+  switch t {
+    case (#leaf) 0;
+    case (#node(_, _, l, r)) {
+           size(l) + size(r)
+         };
+  }
+};
+*/
+
+/*
+ To do: Check invariants for tests:
+
+ isRedBlack =def=
+
+ binarySearchOrder( )
+ and
+ noRedRed( )
+ and
+ okBlackHeight( )
+*/
+
+
+func removeRec<X, Y>(x:X, compareTo:(X, X) -> Comp, t:Tree<X, Y>)
+  : (?Y, Tree<X, Y>)
+{
+  switch t {
+  case (#leaf) { (null, #leaf) };
+  case (#node(c, l, xy, r)) {
+         switch (compareTo(x, xy.0)) {
+         case (#lt) {
+                let (yo, l2) = removeRec(x, compareTo, l);
+                (yo, #node(c, l2, xy, r))
+              };
+         case (#eq) {
+                (xy.1, #node(c, l, (x, null), r))
+              };
+         case (#gt) {
+                let (yo, r2) = removeRec(x, compareTo, r);
+                (yo, #node(c, l, xy, r2))
+              };
+         }
+       }
+  }
+};
+
+
+
+func bal<X, Y>(color:Color, lt:Tree<X, Y>, kv:(X, ?Y), rt:Tree<X, Y>) : Tree<X, Y> {
+  // thank you, algebraic pattern matching!
+  // following notes from [Ravi Chugh](https://www.classes.cs.uchicago.edu/archive/2019/spring/22300-1/lectures/RedBlackTrees/index.html)
+  switch (color, lt, kv, rt) {
+  case (#B, #node(#R, #node(#R, a, x, b), y, c), z, d) #node(#R, #node(#B, a, x, b), y, #node(#B, c, z, d));
+  case (#B, #node(#R, a, x, #node(#R, b, y, c)), z, d) #node(#R, #node(#B, a, x, b), y, #node(#B, c, z, d));
+  case (#B, a, x, #node(#R, #node(#R, b, y, c), z, d)) #node(#R, #node(#B, a, x, b), y, #node(#B, c, z, d));
+  case (#B, a, x, #node(#R, b, y, #node(#R, c, z, d))) #node(#R, #node(#B, a, x, b), y, #node(#B, c, z, d));
+  case _ { #node(color, lt, kv, rt) };
+  }
```

Comment:
@crusso Just for fun: See above for my most extreme use of Motoko pattern-matching to date.  Stolen from Ravi.